Jenkinsfile

```diff
@@ -0,0 +1,259 @@
+
+def config = {
+    owner = 'ksql'
+    slackChannel = '#ksql-alerts'
+    ksql_db_version = "0.6.0-SNAPSHOT"
+    cp_version = "5.5.0-SNAPSHOT" //After done testing this needs to be a beta version.
+    dockerRegistry = '368821881613.dkr.ecr.us-west-2.amazonaws.com/'
+    dockerRepos = ['confluentinc/ksql-cli', 'confluentinc/ksql-rest-app']
+    nodeLabel = 'docker-oraclejdk8-compose-swarm'
+    dockerScan = true
+    cron = '@nightly'
+}
+
+def defaultParams = [
+    booleanParam(name: 'RELEASE_BUILD',
+        description: 'Is this a release build.',
+        defaultValue: false),
+    string(name: 'GIT_REVISION',
+        description: 'The git SHA to create the release build from.',
+        defaultValue: ''),
+    booleanParam(name: 'PROMOTE_TO_PRODUCTION',
+        defaultValue: false,
+        description: 'Promote images to production (DockerHub) for release build images only.'),
+    string(name: 'KSQLDB_VERSION',
+        defaultValue: '',
+        description: 'KSQLDB version to promote to production.'),
+    string(name: 'IMAGE_REVISION',
+        defaultValue: '1',
+        description: 'Revision for Docker images. This is used with PROMOTE_TO_PRODUCTION only'),
+    booleanParam(name: 'UPDATE_LATEST_TAG',
+        defaultValue: false,
+        description: 'Should the latest tag on docker hub be updated to point at this new image version.')
+]
+
+def updateConfig = { c ->
+    c.properties = [parameters(defaultParams)]
+}
+
+def finalConfig = jobConfig(config, [:], updateConfig)
+
+def job = {
+    if (params.RELEASE_BUILD && params.PROMOTE_TO_PRODUCTION) {
+        currentBuild.result = 'ABORTED'
+        error("You can not do a release build and promote to production at the same time.")
+    }
+
+    if (params.RELEASE_BUILD && params.GIT_REVISION == '') {
+        currentBuild.result = 'ABORTED'
+        error("You must provide the GIT_REVISION when doing a release build.")
+    }
+
+    if (params.PROMOTE_TO_PRODUCTION && params.KSQLDB_VERSION == '') {
+        currentBuild.result = 'ABORTED'
+        error("You must provide the KSQLDB_VERSION when promoting images to docker hub.")
+    }
+
+    config.release = params.RELEASE_BUILD
+
+    if (config.release) {
+        // For a release build check out the provided git sha instead of the master branch.
+        if (config.revision == '') {
+            error("If you are doing a release build you must provide a git sha.")
+        }
+        
+        config.revision = params.GIT_REVISION
+        // For a release build we remove the -SNAPSHOT from the version.
+        config.ksql_db_version = config.ksql_db_version.tokenize("-")[0]
```

Comment:
should we resolve `params.KSQL_DB_VERSION` and `config.ksql_db_version`? is one the artifact version and the other the docker version? If so, we probably shouldn't let them diverge

Reply:
@agavra They are doing different things. params.KSQL_DB_VERSION is only used when promoting to production. This is the image version that will get pulled down from ECR, tagged with that version, and then pushed to docker hub. config.ksql_db_version is the version that is specified in the jenkinsfile and gets used for git tagging and also injected into the pom file. Sometimes it will be a snapshot version, but for release builds it will not be. It seems likely to me, although maybe it isn't the case that the current version being built is not the version that you will want to release. For instance after you do a release build of 0.6.0 you will want to change the version to 0.7.0 so the latest builds / snapshots are of that version. Then it could be some time before you are done testing the 0.6.0 release build and ready to promote it. So when you run the build to promote it you need to be able to provide a version to promote, 0.6.0, that is not the same as the current version being build, 0.7.0

Reply:
okay that makes sense - thanks for the explanation